File: backend/app/models/dossier.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Dossier(Base):
    __tablename__ = "dossiers"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)

    client_id: Mapped[int] = mapped_column(
        ForeignKey("clients.id"),
        nullable=False,
        index=True,
    )

    collaborateur_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id"),
        nullable=True,
        index=True,
    )

    titre: Mapped[str] = mapped_column(
        String(200),
        nullable=False,
    )

    type_dossier: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    statut: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        default="En cours",
    )

    date_ouverture: Mapped[datetime] = mapped_column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
    )

    date_cloture: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
    )

    priorite: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="Normale",
    )

    actif: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
    )

    documents = relationship(
        "Document",
        back_populates="dossier",
        cascade="all, delete-orphan",
    )

    taches = relationship(
        "Tache",
        back_populates="dossier",
        cascade="all, delete-orphan",
    )

    client = relationship(
        "Client",
        back_populates="dossiers",
    )

    collaborateur = relationship(
        "User",
        foreign_keys=[collaborateur_id],
    )

    factures = relationship(
        "Facture",
        back_populates="dossier",
    )

File: frontend/lib/models/dossier.dart
class Dossier {
  final int id;
  final int clientId;
  final int? collaborateurId;
  final String titre;
  final String typeDossier;
  final String? description;
  final String statut;
  final String priorite;
  final DateTime dateOuverture;
  final DateTime? dateCloture;
  final bool actif;

  const Dossier({
    required this.id,
    required this.clientId,
    this.collaborateurId,
    required this.titre,
    required this.typeDossier,
    this.description,
    required this.statut,
    required this.priorite,
    required this.dateOuverture,
    this.dateCloture,
    required this.actif,
  });

  factory Dossier.fromJson(Map<String, dynamic> json) {
    return Dossier(
      id: json['id'] as int,
      clientId: json['client_id'] as int,
      collaborateurId: (json['collaborateur_id'] as num?)?.toInt(),
      titre: json['titre'] as String,
      typeDossier: json['type_dossier'] as String,
      description: json['description'] as String?,
      statut: json['statut'] as String? ?? 'En cours',
      priorite: json['priorite'] as String? ?? 'Normale',
      dateOuverture: DateTime.parse(json['date_ouverture'] as String),
      dateCloture: json['date_cloture'] != null
          ? DateTime.parse(json['date_cloture'] as String)
          : null,
      actif: json['actif'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'client_id': clientId,
      'collaborateur_id': collaborateurId,
      'titre': titre,
      'type_dossier': typeDossier,
      'description': description,
      'statut': statut,
      'priorite': priorite,
      'date_ouverture': dateOuverture.toIso8601String(),
      'date_cloture': dateCloture?.toIso8601String(),
      'actif': actif,
    };
  }
}
